experimental/Schemes/AffineSchemes.jl

```diff
@@ -999,19 +1001,23 @@ end
 end
 
 @attr function dim(X::AbsSpec{<:Ring, <:MPolyLocalizedRing})
-  return ngens(ambient_ring(X))
+  # the following line is supposed to refer the problem to the 
+  # algebra side of the problem 
+  return dim(ideal(ambient_ring(X), [zero(ambient_ring(X))]))
 end
 
 @attr function dim(X::AbsSpec{<:Ring, <:MPolyRing})
-  return ngens(ambient_ring(X))
+  return dim(ideal(ambient_ring(X), [zero(ambient_ring(X))]))
 end
 
 @attr function dim(X::AbsSpec{<:Ring, <:MPolyQuo})
   return dim(modulus(OO(X)))
 end
 
-function codim(X::Spec)
-  return ngens(base_ring(OO(X)))-dim(X)
+# TODO: This is not 100% mathematically correct, since codimension is 
+# a local concept. Should we therefore forbid this method?
+@attr function codim(X::Spec)
```

Comment:
You are actually aiming at height(modulus(X)), right? This involves a primary (or equidimensional) decomposition in the general, i.e. non-equidimensional, case and should be referred to the algebraic side, where all the functionality should be available.

Reply:
Yes, `height(modulus(OO(X))` would be one candidate. It depends what this function will be used for. This has probably been introduced to determine the expected rank of the jacobian matrix of some set of generators of `modulus(OO(X))`. In that case, invoking a primary decomposition first does not seem reasonable. So I would really vote for deleting this function and leave it to the user to exctract the information they need for their particular applications.

Reply:
Why would you forbid this? It is well-defined globally and can be computed without primary decomposition, the hightest-dimensional component wins.

Reply:
Yes, you are right, of course the highest-dimension component wins (Wald vor Bäumen nicht gesehen....). 
This should just be 
                    dimension of ambient scheme - dimension of X.

By the way: You do not get as much sensible information from the Jacobian criterion in the arithmetic case anyway...